### PR TITLE
bpo-42213: Remove redundant cyclic GC hack in sqlite3

### PR DESCRIPTION
**Pull request by erlend-aasland**
The sqlite3 module now fully implements the GC protocol; there's no
need for this workaround anymore.

https://bugs.python.org/issue42213

### CODE REVIEW

**Comment by erlend-aasland** [comment]
Pablo/Victor: Whenever you have time 🙏🏻

**Comment by erlend-aasland** [comment]
I'm not sure if this should be backported to 3.10. It's just a cleanup, so I guess it's fine to backport it, but on the other hand there is no harm in not backporting it. I have no strong preference here.

**Comment by erlend-aasland** [comment]
The Windows failures looks very similar to the GC issues in #24203.
The macOS failure is a spurious ssl failure.

**Comment by erlend-aasland** [comment]
Adding 209196eaaa44d02551dad00516ff32fb447c1d8f fixes the test issues on Windows.

File: Lib/sqlite3/test/dbapi.py [hunk 1]
```diff
@@ -190,6 +193,8 @@ def test_open_uri(self):
         with sqlite.connect('file:' + TESTFN + '?mode=ro', uri=True) as cx:
             with self.assertRaises(sqlite.OperationalError):
                 cx.execute('insert into test(id) values(1)')
+        cx.close()
+        gc_collect()
```
Comment (pablogsal) [inline Lib/sqlite3/test/dbapi.py]: Why the falls to GC collect here? These should not be necessary

Comment (erlend-aasland) [reply Lib/sqlite3/test/dbapi.py]: I know; it's the same issue as in #24203. I'll try my best to reproduce this on my Mac.

Comment (erlend-aasland) [reply Lib/sqlite3/test/dbapi.py]: FYI, discovered #26475 while debugging this, but it does unfortunately not fix this issue.

**Comment by erlend-aasland** [comment]
FYI, rebased onto `main`.

**Comment by vstinner** [comment]
Did you run `./python -m test test_sqlite -R 3:3` to check for reference leaks?

**Comment by vstinner** [comment]
Windows (x64) CI job failed:

```

2021-06-01T11:02:07.4501179Z ======================================================================
2021-06-01T11:02:07.4503272Z ERROR: test_open_uri (sqlite3.test.dbapi.ConnectionTests)
2021-06-01T11:02:07.4581189Z ----------------------------------------------------------------------
2021-06-01T11:02:07.4598093Z Traceback (most recent call last):
2021-06-01T11:02:07.4600441Z   File "D:\a\cpython\cpython\lib\test\support\os_helper.py", line 238, in unlink
2021-06-01T11:02:07.4633846Z     _unlink(filename)
2021-06-01T11:02:07.4636229Z   File "D:\a\cpython\cpython\lib\test\support\os_helper.py", line 278, in _unlink
2021-06-01T11:02:07.4637841Z     _waitfor(os.unlink, filename)
2021-06-01T11:02:07.4639071Z   File "D:\a\cpython\cpython\lib\test\support\os_helper.py", line 246, in _waitfor
2021-06-01T11:02:07.4640337Z     func(pathname)
2021-06-01T11:02:07.4642275Z PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: '@test_1508_tmp�'
2021-06-01T11:02:07.4644053Z 
2021-06-01T11:02:07.4645203Z ======================================================================
2021-06-01T11:02:07.4646483Z ERROR: test_open_with_path_like_object (sqlite3.test.dbapi.ConnectionTests)
2021-06-01T11:02:07.4648280Z Checks that we can successfully connect to a database using an object that
2021-06-01T11:02:07.4650124Z ----------------------------------------------------------------------
2021-06-01T11:02:07.4652596Z Traceback (most recent call last):
2021-06-01T11:02:07.4655019Z   File "D:\a\cpython\cpython\lib\sqlite3\test\dbapi.py", line 182, in test_open_with_path_like_object
2021-06-01T11:02:07.4670040Z     cx.execute('create table test(id integer)')
2021-06-01T11:02:07.4672512Z sqlite3.OperationalError: table test already exists
2021-06-01T11:02:07.4688676Z 
2021-06-01T11:02:07.4708289Z ======================================================================
2021-06-01T11:02:07.4724689Z ERROR: test_open_with_path_like_object (sqlite3.test.dbapi.ConnectionTests)
2021-06-01T11:02:07.4728000Z Checks that we can successfully connect to a database using an object that
2021-06-01T11:02:07.4811284Z ----------------------------------------------------------------------
2021-06-01T11:02:07.4831832Z Traceback (most recent call last):
2021-06-01T11:02:07.4834116Z   File "D:\a\cpython\cpython\lib\test\support\os_helper.py", line 238, in unlink
2021-06-01T11:02:07.4864854Z     _unlink(filename)
2021-06-01T11:02:07.4867085Z   File "D:\a\cpython\cpython\lib\test\support\os_helper.py", line 278, in _unlink
2021-06-01T11:02:07.4868728Z     _waitfor(os.unlink, filename)
2021-06-01T11:02:07.4870312Z   File "D:\a\cpython\cpython\lib\test\support\os_helper.py", line 246, in _waitfor
2021-06-01T11:02:07.4871577Z     func(pathname)
2021-06-01T11:02:07.4873516Z PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: '@test_1508_tmp�'
2021-06-01T11:02:07.4875042Z 
2021-06-01T11:02:07.4876184Z ======================================================================
2021-06-01T11:02:07.4877721Z ERROR: test_trace_callback_content (sqlite3.test.hooks.TraceCallbackTests)
2021-06-01T11:02:07.4879746Z ----------------------------------------------------------------------
2021-06-01T11:02:07.4881969Z Traceback (most recent call last):
2021-06-01T11:02:07.4884465Z   File "D:\a\cpython\cpython\lib\test\support\os_helper.py", line 238, in unlink
2021-06-01T11:02:07.4898652Z     _unlink(filename)
2021-06-01T11:02:07.4900840Z   File "D:\a\cpython\cpython\lib\test\support\os_helper.py", line 278, in _unlink
2021-06-01T11:02:07.4931605Z     _waitfor(os.unlink, filename)
2021-06-01T11:02:07.4947677Z   File "D:\a\cpython\cpython\lib\test\support\os_helper.py", line 246, in _waitfor
2021-06-01T11:02:07.4949998Z     func(pathname)
2021-06-01T11:02:07.5029259Z PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: '@test_1508_tmp�'
2021-06-01T11:02:07.5046117Z 
2021-06-01T11:02:07.5047969Z ----------------------------------------------------------------------
```

**Comment by vstinner** [comment]
Ah, there is also a warning on Windows (same CI job):

```
2021-06-01T11:11:08.6974894Z Warning -- files was modified by test_sqlite
2021-06-01T11:11:08.6975521Z   Before: []
2021-06-01T11:11:08.6977644Z   After:  ['@test_1748_tmp�'] 
```

**Comment by erlend-aasland** [comment]
> Did you run `./python -m test test_sqlite -R 3:3` to check for reference leaks?
> 
> 

Yes, I did :)

**Comment by erlend-aasland** [comment]
> Windows (x64) CI job failed:

Yes, this is the reason for the draft/WIP status. On Windows, the GC uses a lot more time to break the ref. cycle between the statement cache and the connection. From what I can see, it seems that it just loops longer; there's a load of traverse calls before the cycle is broken. On Mac and Linux, the cycle is broken much faster. This results in the objects living slightly longer on Windows, so sqlite3 is clinging on to `TESTFN`, preventing `unlink` from removing it (as seen in the first traceback). Calling unlink repeatedly until it succeeds works, because eventually GC manages to clean up, and sqlite3 releases the file. It seems that the GC just behaves differently on Windows.

**Comment by erlend-aasland** [comment]
> In general, I dislike relying on implicit resource management. I would prefer to emit a ResourceWarning if a resource is not released explicitly (by calling a close() method for example). Maybe some tests should call the close() method explicitly?

That's how I "solved" this in #24203: see d2078bced6179882fe911be091b467e506a0db56.

It's exactly the same problem:

The LRU statement cache (`pysqlite_Connection.statement_cache`) has a "back-reference" to `sqlite3.Connection` (`pysqlite_Cache.factory`), because statement creation happens in `sqlite3.Connection.__call__`. I find that design strange and unfortunate.

Previously (or should I say currently), `sqlite3.Connection.__init__` decrefs `self` after creating the LRU cache, and then tells `sqlite3.Cache` to _not_ decref the statement factory (`pysqlite_Cache.factory`) when the cache is deallocated. In practice the cache is using a borrowed ref. to the connection.

**Comment by vstinner** [comment]
Keeping Python objects in memory is fine. What is not fine is to hold an OS release, like a file. Do you mean that calling close() does not close the file on disk?

**Comment by erlend-aasland** [comment]
> Keeping Python objects in memory is fine. What is not fine is to hold an OS release, like a file. Do you mean that calling close() does not close the file on disk?

It does close it, but not until GC is done.

**Comment by erlend-aasland** [comment]
> `with sqlite.connect(path) as cx:` doesn't close the connection at the end of the with block? Is it a deliberate behavior?

The context manager decref's the connection object, which calls dealloc, which closes the conneciton. _But_, arriving at connection dealloc takes a lot of time because of the garbage collector using a lot of time to break the ref. cycle. Maybe I'm terrible at explaining this, but what's happening (as I see it) is:

1. `__enter__`
2. do stuff
3. `__exit__`
4. traverse loop (ref. cycle)
5. unlink => fails, because GC is still traversing
6. traverse done => connection dealloc/clean => sqlite3 finally releases `TESTFN`, but it's too late

My computer time today is very fragmented, so my replies will be short and hopefully not too messy :)

**Comment by vstinner** [comment]
> __exit__ does not close the connection, it only tries to commit or rollback what happened inside it

Well, again, IMO sqlite3 must behave as the io module: emit a ResourceWarning if a connection is not closed **explicitly**.

All tests must explicitly close the connection.

Closing a file must not depend if a sqlite object is part of a reference cycle or not. There are too many ways to create reference cycles in Python.

**Comment by erlend-aasland** [comment]
> Well, again, IMO sqlite3 must behave as the io module: emit a ResourceWarning if a connection is not closed **explicitly**.
> 
> All tests must explicitly close the connection.
> 
> Closing a file must not depend if a sqlite object is part of a reference cycle or not. There are too many ways to create reference cycles in Python.

In that case, I believe explicitly closing the connection and explicitly triggering gc.collect is the best thing to do here.

**Comment by erlend-aasland** [comment]
@vstinner, what do you think of the workarounds I've added? Are they acceptable, or do you want me to add ResourceWarning?

**Comment by vstinner** [comment]
> In that case, I believe explicitly closing the connection and explicitly triggering gc.collect is the best thing to do here.

I don't understand why calling close() is not enough to close the DB file. I don't understand how the fact that a sqlite object is involved in a reference cycle prevents to close the file.

Example:

```
>>> import os
>>> f=open("/etc/issue")
# open file
>>> fd=f.fileno()
>>> os.fstat(fd)
os.stat_result(st_mode=33188, st_ino=19136063, st_dev=35, st_nlink=1, st_uid=0, st_gid=0, st_size=28, st_atime=1622620921, st_mtime=1618252698, st_ctime=1619553001)

# closed file
>>> f.close()
>>> os.fstat(fd)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [Errno 9] Bad file descriptor
```

Even if the `f` Python object is not garbage collected yet, the file descriptor is closed and fstat(fd) fails with EBADF, as expected.

**Comment by vstinner** [comment]
I'm not sure that we understand each other.

Calling explicity close() and emitting ResourceWarning is one thing.

Reducing the risk of reference cycle is another thing. IMO it's unrelated.

```
with sqlite.connect(path) as cx:
  conn = cx # "reference cycle"
cx.close()
# I expect the DB to be closed at that point even if 'conn' object still exists
```

File: Lib/sqlite3/test/dbapi.py [hunk 2]
```diff
@@ -170,10 +171,16 @@ def test_in_transaction_ro(self):
         with self.assertRaises(AttributeError):
             self.cx.in_transaction = True
 
+
+class OpenTests(unittest.TestCase):
+    def tearDown(self):
+        # bpo-42213: ensure that TESTFN is closed before unlinking
+        gc_collect()
```
Comment (vstinner) [inline Lib/sqlite3/test/dbapi.py]: I am against a PR which calls gc.collect() to ensure that a resource is released. It doesn't look reliable at all. It sounds like a way to hide a real bug, and it will likely bite us later.

Comment (erlend-aasland) [reply Lib/sqlite3/test/dbapi.py]: I agree; I'll revert the workaround.

**Comment by erlend-aasland** [comment]
> I'm not sure that we understand each other.

We're probably talking past each other, yes.

> I don't understand why calling close() is not enough to close the DB file. I don't understand how the fact that a sqlite object is involved in a reference cycle prevents to close the file.

Clarification: We don't call `close()` as in the syscall `close`; we call the SQLite C API `sqlite3_close_v2` (http://sqlite.org/c3ref/close.html):

_"If sqlite3_close_v2() is called with unfinalized prepared statements, unclosed BLOB handlers, and/or unfinished sqlite3_backups, it returns SQLITE_OK regardless, but instead of deallocating the database connection immediately, it marks the database connection as an unusable "zombie" and makes arrangements to automatically deallocate the database connection after all prepared statements are finalized, all BLOB handles are closed, and all backups have finished. The sqlite3_close_v2() interface is intended for use with host languages that are garbage collected, and where the order in which destructors are called is arbitrary."_

I've tried changing this to `sqlite3_close` and checking that the return value is not `SQLITE_BUSY`, but the problem persists.

**Comment by pablogsal** [comment]
I think there is some confusion between @vstinner and @erlend-aasland. Let me try to clarify what are the expectations so is easier to reconduct the discussion:

* It seems that the object calls close **on destruction**.
* Being in a `gc` cycle prevents destruction.
* The context manager doesn't close the connection.

One of the things that @vstinner is saying (and I very much agree with) is that we should never depend on calling `gc.collect()` to clean resources in the test suite. Is very unreliable and is almost always hiding bugs. For instance, in this case as the connexion is closed on destruction, is preferable to schedule a cleanup method (calling `close()` on the connection) but not relying on it being closed on destruction, as that is *implicit* resource management, which is bad.

Furthermore, @vstinner thinks that this implicit resource management (closing on destruction) is so bad that a warning should be emmited if we ever reach that code with the connection open, and all users should explicitly call `close()` or alternative we should change the context manager to close the connection, which is probably backwards compact.

In short:

* Don't use `gc.collect()` in the test suite to do cleanups.
* We should think how to ensure the connection is closed using explicit resource management.

**Comment by vstinner** [comment]
I understand that it is possible that the database file remains open after sqlite3_close_v2() is called, if some other sqlite objects still exist. Oh, this is counter intuitive. How can a programmer know if a database file is closed or not? Try to delete it on Windows? :-)

Would it help to emit a warning or even an hard exception if this case happen?

**Comment by erlend-aasland** [comment]
Not the Spanish inquisition!!! 😂 Thanks, @pablogsal! :) I believe we're getting closer to understanding the various issues.

Victor:
> Oh, this is counter intuitive. How can a programmer know if a database file is closed or not? Try to delete it on Windows? :-)

Yes, it is counter-intuitive, and no, there's no API for detecting if a database file is closed (AFAIK). `fstat`? :)

**Comment by vstinner** [comment]
I understand that the close() method must destroy/release/close all sqlite objects **before** calling sqlite3_close_v2().

Said differently, would it be possible to implement a public or private close() method on each Python sqlite object which somehow prevents to close a database (sqlite3_close_v2), to release the resources, without having to destroy the Python object?

See my file object: even if the Python object is not destroyed, the inner resource is released. It requires all file methods to raise an exception if the file is closed.

**Comment by erlend-aasland** [comment]
Anyway, I must take a break now; back at the computer in 5/6 hours :)

**Comment by vstinner** [comment]
> Anyway, I must take a break now; back at the computer in 5/6 hours :)

The Python project is developed asynchronously. You don't *have to* reply in less than 1 hour. It's perfectly fine to take 1 week or even 1 month to reply to a review. Your message sends notifications which are counter-productive. I go to the PR to see the comment... to say that you will reply later, ah. On IRC, I only get a notification that you wrote a comment, but I don't get the content.

**Comment by erlend-aasland** [comment]
Looks like I got it right. PTAL, @vstinner & @pablogsal. Thanks for your patience and guidance.

File: Modules/_sqlite/connection.c [hunk 3]
```diff
@@ -353,23 +353,18 @@ static PyObject *
...
 
-    pysqlite_do_all_statements(self, ACTION_FINALIZE, 1);
-
     if (self->db) {
-        rc = sqlite3_close_v2(self->db);
+        /* Free pending statements before closing. This implies also cleaning
+         * up cursors, as they may have strong refs to statements. */
+        Py_CLEAR(self->statement_cache);
+        Py_CLEAR(self->statements);
+        Py_CLEAR(self->cursors);
```
Comment (pablogsal) [inline Modules/_sqlite/connection.c:365]: What happens if you use the connection after you call `close()`? Are we checking for NULL for these veriables everywhere else? (Cannot check myself as I'm on my phone)

Comment (erlend-aasland) [reply Modules/_sqlite/connection.c:365]: You'll get a `sqlite3.ProgrammingError`. There's a lot of sanity checks in the code for this. The sqlite3 test suite runs fine, but of course, we do not have 100% code coverage yet. (Digression: see [issue 43553](https://bugs.python.org/issue43553) for improving `sqlite3` code coverage.)

After close, if you try to use a cursor, fetch from a pending statement, or manipulate the connection, you'll get a `sqlite3.ProgrammingError`:
```
>>> cx = sqlite3.connect(":memory:")
>>> cu = cx.cursor()
>>> res = cu.execute("select 1")
>>> cx.close()
>>> res.fetchall()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
sqlite3.ProgrammingError: Cannot operate on a closed database.
>>> cu.execute("select 1")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
sqlite3.ProgrammingError: Cannot operate on a closed database.
>>> cx.create_function("test", 1, lambda x: x)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
sqlite3.ProgrammingError: Cannot operate on a closed database.
>>> 
```

**Comment by vstinner** [comment]
`_sqlite3.Connection.__enter__()` doesn't call pysqlite_check_connection() (check if the DB is open). IMO it should.

The following methods check *indirectly* if the connection is closed.

*_sqlite3.Connection.execute(): call self.cursor().execute()
* _sqlite3.Connection.executemany(): call self.cursor().executemany()
* _sqlite3.Connection.executescript(): call self.cursor().executescript()
* `_sqlite3.Connection.__exit__()`: call commit() or rollback()

Other methods call pysqlite_check_connection().

File: Lib/sqlite3/test/dbapi.py [hunk 4]
```diff
@@ -170,26 +170,35 @@ def test_in_transaction_ro(self):
...
         """ Checks that we can successfully connect to a database using an object that
             is PathLike, i.e. has __fspath__(). """
-        self.addCleanup(unlink, TESTFN)
         class Path:
             def __fspath__(self):
                 return TESTFN
         path = Path()
         with sqlite.connect(path) as cx:
             cx.execute('create table test(id integer)')
+        cx.close()
```
Comment (vstinner) [inline Lib/sqlite3/test/dbapi.py]: Maybe a helper context manager would be helpful? It would close the DB and then remove the file. Something like:

```
@contextlib.contextmanager
def connect(path):
  conn = sqlite.connect(path)
  try:
    yield conn
  finally:
    conn.close()
    unlink(path)
```

I'm not sure if it's worth it if there are just a few users of such helper function. Your PR is also good like that.

File: Lib/sqlite3/test/hooks.py [hunk 5]
```diff
@@ -259,6 +259,8 @@ def trace(statement):
         cur.execute(queries[0])
         con2.execute("create table bar(x)")
         cur.execute(queries[1])
+        con1.close()
+        con2.close()
```
Comment (vstinner) [inline Lib/sqlite3/test/hooks.py]: May use try/finally to ensure that the connection is closed if an assertion fails. Or use an helper function.

**Review by vstinner — COMMENTED** [review]
Hum. Can you try to split your PR in two parts? First PR to rewrite the C code, second PR to rewrite the tests.

File: Lib/sqlite3/test/hooks.py (hunk 5 above)
Comment (erlend-aasland) [reply Lib/sqlite3/test/hooks.py]: Good point

File: Lib/sqlite3/test/dbapi.py (hunk 4 above)
Comment (erlend-aasland) [reply Lib/sqlite3/test/dbapi.py]: I'm not sure it's worth it, but I'll keep it in mind if we add more such users.

File: Lib/sqlite3/test/hooks.py (hunk 5 above)
Comment (erlend-aasland) [reply Lib/sqlite3/test/hooks.py]: I added d31403eb9885eb4654741a84b28f5dcd7179acad.

File: Modules/_sqlite/connection.c (hunk 3 above)
Comment (erlend-aasland) [reply Modules/_sqlite/connection.c:365]: I added some more tests wrt. this: 1b23df16d0dfad14b3f401ccbeeb34b7cbf3b82e
Also, there's some regression tests that exercise operations on closed connections (and closed cursors).

Let me know if you want me to add more tests.

**Comment by erlend-aasland** [comment]
> `_sqlite3.Connection.__enter__()` doesn't call pysqlite_check_connection() (check if the DB is open). IMO it should.

Fixed in 7d8014ab5b3672ee4994abecb7bc63578e56cde5.

> Hum. Can you try to split your PR in two parts? First PR to rewrite the C code, second PR to rewrite the tests.

Yes, I could do that, but that would make the CI fail on the first PR.

File: Lib/sqlite3/test/dbapi.py (hunk 4 above)
Comment (erlend-aasland) [reply Lib/sqlite3/test/dbapi.py]: I changed my mind; the helper is nice. I'll add it :)

Comment (erlend-aasland) [reply Lib/sqlite3/test/dbapi.py]: See aba8dd724618df1a3ab2cad90b03436cb9bc9fd8

**Comment by vstinner** [comment]
> Yes, I could do that, but that would make the CI fail on the first PR.

Hum, I am not strictly thinking about C vs Python, but more something like that:

First PR:

* Add Py_CLEAR in pysqlite_connection_close_impl
* Add connection_close()
* pysqlite_connection_enter_impl() change
* unit tests for closed connections/cursors

Second PR:

* Remove decref_factory
* Remove gc.collect() calls
* Add managed_connect() in test

**Comment by vstinner** [comment]
This PR is no longer only about "Remove redundant cyclic GC hack in sqlite3" and "The sqlite3 module now fully implements the GC protocol; there's no need for this workaround anymore." It's now way more than that.

**Comment by erlend-aasland** [comment]
True. Can I use the same issue number?

**Comment by erlend-aasland** [comment]
I'll add an issue for hardening close and direct the first PR towards that.

**Comment by vstinner** [comment]
> Can I use the same issue number?

Yes. It's a good practice to put related changes in the same bpo.

File: Modules/_sqlite/connection.c (hunk 3 above)
Comment (erlend-aasland) [reply Modules/_sqlite/connection.c:365]: Here's a strange observation:
I did a ref leak test on Windows yesterday, and it segfaulted on `test_bpo31770`. After some testing, I reverted the change highlighted in this conversation; I removed the three `Py_CLEAR`, added `pysqlite_do_all_statements(self, ACTION_FINALIZE, 1)` back, but kept `connection_close()`. That fixed the ref leak test on Windows, and the test suite still worked; it did not "leak" open test database files. So, for the new PR, I'll keep this modification. We'll see how the CI fares :)

Comment (erlend-aasland) [reply Modules/_sqlite/connection.c:365]: cc. @vstinner

Comment (erlend-aasland) [reply Modules/_sqlite/connection.c:365]: CI seems to be fine with this: https://github.com/erlend-aasland/cpython/pull/9
